pkg/generic/thrift/json_go116plus_amd64.go

```diff
@@ -0,0 +1,178 @@
+	if svc.DynamicgoDsc != nil {
+		dty = svc.DynamicgoDsc.Functions()[method].Request()
+		if !isClient {
+			dty = svc.DynamicgoDsc.Functions()[method].Response()
+		}
+		if dty == nil {
+			return nil, perrors.NewProtocolErrorWithMsg("dynamicgo TypeDescriptor is nil")
+		}
+	}
+	ws := &WriteJSON{
+		ty:             ty,
+		dty:            dty,
+		hasRequestBase: fnDsc.HasRequestBase && isClient,
+		isClient:       isClient,
+	}
+	return ws, nil
+}
+
+// Write write json string to out thrift.TProtocol
+func (m *WriteJSON) Write(ctx context.Context, out thrift.TProtocol, msg interface{}, requestBase *Base) error {
+	if !m.hasRequestBase {
+		requestBase = nil
+	}
+	if requestBase != nil {
+		m.dyOpts.EnableThriftBase = true
+	}
+
+	// msg is void
+	if _, ok := msg.(descriptor.Void); ok {
+		if err := m.writeHead(out); err != nil {
+			return err
+		}
+		if err := m.writeFields(ctx, out, Void, nil, nil, nil); err != nil {
+			return err
+		}
+		return writeTail(out)
+	}
+
+	// msg is string
+	s, ok := msg.(string)
+	if !ok {
+		return perrors.NewProtocolErrorWithType(perrors.InvalidData, "decode msg failed, is not string")
+	}
+
+	cv := j2t.NewBinaryConv(*m.dyOpts)
+	transBuff := utils.StringToSliceByte(s)
+	dbuf := mcache.Malloc(len(transBuff))[0:0]
+
+	if err := m.writeHead(out); err != nil {
+		return err
+	}
+	if err := m.writeFields(ctx, out, String, &cv, transBuff, dbuf); err != nil {
+		return err
+	}
+	return writeTail(out)
+}
+
+type MsgType int
+
+const (
+	Void MsgType = iota
+	String
+)
+
+func (m *WriteJSON) writeFields(ctx context.Context, out thrift.TProtocol, msgType MsgType, cv *j2t.BinaryConv, transBuff, dbuf []byte) error {
+	for _, field := range m.dty.Struct().Fields() {
+		// Exception field
+		if !m.isClient && field.ID() != 0 {
+			// generic server ignore the exception, because no description for exception
+			// generic handler just return error
+			continue
+		}
+
+		if err := out.WriteFieldBegin(field.Name(), field.Type().Type().ToThriftTType(), int16(field.ID())); err != nil {
+			return err
+		}
+		switch msgType {
+		case Void:
+			if err := out.WriteStructBegin(field.Name()); err != nil {
+				return err
+			}
+			if err := out.WriteFieldStop(); err != nil {
+				return err
+			}
+			if err := out.WriteStructEnd(); err != nil {
+				return err
+			}
+		case String:
+			// encode using dynamicgo
+			// json []byte to thrift []byte
+			if err := cv.DoInto(ctx, field.Type(), transBuff, &dbuf); err != nil {
+				return err
+			}
+		}
+		// WriteFieldEnd has no content
+		// if err := out.WriteFieldEnd(); err != nil {
+		// 	return err
+		// }
+	}
+	if msgType == String {
+		tProt, ok := out.(*cthrift.BinaryProtocol)
+		if !ok {
+			return perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+		}
+		buf, err := tProt.ByteBuffer().Malloc(len(dbuf))
+		if err != nil {
+			return err
+		} else {
+			// TODO: implement MallocAck() to achieve zero copy
+			copy(buf, dbuf)
+		}
+		mcache.Free(dbuf)
```

Comment:
move this line below `dbuf := mcache.Malloc(len(transBuff))[0:0]` and use defer to delay execution.

Reply:
fixed